File: Binary Search Tree/449. Serialize and Deserialize BST.cpp
/*
449. Serialize and Deserialize BST

Serialization is the process of converting a data structure or object into a sequence of bits 
so that it can be stored in a file or memory buffer, 
or transmitted across a network connection link to be reconstructed later in the same or another computer environment.

Design an algorithm to serialize and deserialize a binary search tree. 
There is no restriction on how your serialization/deserialization algorithm should work. 
You just need to ensure that a binary search tree can be serialized to a string and
 this string can be deserialized to the original tree structure.

The encoded string should be as compact as possible.

Note: Do not use class member/global/static variables to store states. Your serialize and deserialize algorithms should be stateless.


*/





// Time:  O(n)
// Space: O(h)


// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));




// Time:  O(n)
// Space: O(h)
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        string outstring;
        serializehelper(root, &outstring);
        return outstring;
    }
    
    void serializehelper(TreeNode* root, string *outstring) {
        if(root){
            *outstring += to_string(root->val)+" ";
            serializehelper(root->left, outstring);
            serializehelper(root->right, outstring);
        }
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        int pos = 0;
        return deserializehelper(numeric_limits<int>::min(), numeric_limits<int>::max(),data,&pos);
    }
    
    TreeNode* deserializehelper(int minval, int maxval, const string & data, int * pos) { 
        //必须需要pointer 或者reference
    /* 
    比如 结果是 
         5
        3 6
       2   7
    data = 5 3 6 2 7， 当左树loop 完，到右面，val=5，但是要保证pos指向index 6, 不能是指向3的
     */
        if(*pos==data.size()) 
            return nullptr;
        int j = data.find(' ', *pos);
        int val = stoi(data.substr(*pos, j - *pos));
        if(val>minval && val<maxval){
            TreeNode * node = new TreeNode(val);
            *pos = j+1;
            node->left = deserializehelper(minval, val, data, pos);
            node->right = deserializehelper(val, maxval, data,pos);
            return node;
        }
        return nullptr;
    }
};


/*
Sharing my solution which doesn't have to parse string for comma at all!

The encoding schema is preorder of BST, and to decode this we can use the same preorder traversal to do it in one pass with recursion in O(n) time.

To minimize the memory, I used binary format instead of ascii format for each integer, just burn those int into 4 chars will save you a lot!!!

Really if using ASCII numbers you are paying a lot of penalty memory for integers over 4 digit long and parsing comma is just as painful.

*/

class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        string order;
        inorderDFS(root, order);
        return order;
    }
    
    inline void inorderDFS(TreeNode* root, string& order) {
        if (!root) return;
        char buf[4];
        memcpy(buf, &(root->val), sizeof(int)); //burn the int into 4 chars
        for (int i=0; i<4; i++) order.push_back(buf[i]);
        inorderDFS(root->left, order);
        inorderDFS(root->right, order);
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        int pos = 0;
        return reconstruct(data, pos, INT_MIN, INT_MAX);
    }
    
    inline TreeNode* reconstruct(const string& buffer, int& pos, int minValue, int maxValue) {
        if (pos >= buffer.size()) return NULL; //using pos to check whether buffer ends is better than using char* directly.
        
        int value;
        memcpy(&value, &buffer[pos], sizeof(int));
        if (value < minValue || value > maxValue) return NULL;
        
        TreeNode* node = new TreeNode(value);
        pos += sizeof(int);
        node->left = reconstruct(buffer, pos, minValue, value);
        node->right = reconstruct(buffer, pos, value, maxValue);
        return node;
    }
};


class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        string outstring;
        serializehelper(root, &outstring);
        return outstring;
    }
    
    void serializehelper(TreeNode* root, string *outstring) {
        if(root){
            char buf[4];
            memcpy(buf,&(root->val),sizeof(int));
            for(int i = 0;i<4; i++)  *outstring += buf[i];
            serializehelper(root->left, outstring);
            serializehelper(root->right, outstring);
        }
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        int pos = 0;
        return deserializehelper(numeric_limits<int>::min(), numeric_limits<int>::max(),data,&pos);
    }
    
    TreeNode* deserializehelper(int minval, int maxval, const string & data, int * pos) {
        if(*pos==data.size()) 
            return nullptr;
        int val;
        memcpy(&val, &data[*pos], sizeof(int));
        if(val>minval && val<maxval){
            TreeNode * node = new TreeNode(val);
            *pos += sizeof(int);
            node->left = deserializehelper(minval, val, data, pos);
            node->right = deserializehelper(val, maxval, data,pos);
            return node;
        }
        return nullptr;
    }
};

File: Binary Search Tree/449. Serialize and Deserialize BST.py
"""
449. Serialize and Deserialize BST
"""
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        self.string = ""
        def preorder(root):
            if root:
                self.string += str(root.val)+" "
                preorder(root.left)
                preorder(root.right)
        preorder(root)
        return self.string

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        self.pos = 0
        self.dic = [int(i) for i in data.split(" ")[:-1]]
        def helper(minval, maxval):
            if self.pos == len(self.dic):
                return None
            val = self.dic[self.pos]
            if val>minval and val<maxval:
                node = TreeNode(val)
                self.pos += 1
                node.left = helper(minval,val)
                node.right = helper(val,maxval)
                return node
            else:
                return None
        return helper(-10**10, 10**10)
        

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))



"""
作弊式回答
"""
class TreeNode(object):
    def __init__(self, x):
        self.val=x
        self.left=None
        self.right=None
            

class Codec:

    def serialize(self, root):
        return root

    def deserialize(self, data):
        return data

"""
用deque 加popleft()
"""

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        string = []
        def preorder(root):
            if root:
                string.append(root.val)
                preorder(root.left)
                preorder(root.right)
        preorder(root)
        return " ".join(map(str,string))

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        dic = collections.deque([int(i) for i in data.split()])
        def helper(minval, maxval):
            if len(dic) == 0:
                return None
            if minval<dic[0] <maxval:
                node = TreeNode(dic.popleft())
                node.left = helper(minval,node.val)
                node.right = helper(node.val,maxval)
                return node
            else:
                return None
        return helper(float('-infinity'), float('infinity'))